src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+  ]
+  incoming_messages = reduce(vcat, incoming_messages; init=ITensor[])
+
+  contract_list = ITensor[
+    incoming_messages
+    Vector{ITensor}(subgraph(ptn, src(edge)))
+  ]
+
+  return contractor(contract_list; contractor_kwargs...)
+end
+
+"""
+Do a sequential update of the message tensors on `edges`
+"""
+function update(
+  bpc::BeliefPropagationCache,
+  edges::Vector{<:PartitionEdge};
+  compute_norm=false,
+  mt_constructor=default_mt_constructor,
+  kwargs...,
+)
+  bpc_updated = copy(bpc)
+  mts = message_tensors(bpc_updated)
+  c = 0
+  for e in edges
+    mts[e] = updated_message_tensor(bpc_updated, e; mt_constructor, kwargs...)
+    if compute_norm
+      c += message_tensor_diff(message_tensor(bpc, e; mt_constructor), mts[e])
+    end
+  end
+  return bpc_updated, c / (length(edges))
+end
+
+"""
+Update the message tensor on a single edge
+"""
+function update(bpc::BeliefPropagationCache, edge::PartitionEdge; kwargs...)
+  return update(bpc, [edge]; kwargs...)
+end
+
+"""
+Do parallel updates between groups of edges of all message tensors
+Currently we send the full message tensor data struct to update for each edge_group. But really we only need the
+mts relevant to that group.
+"""
+function update(
+  bpc::BeliefPropagationCache, edge_groups::Vector{<:Vector{<:PartitionEdge}}; kwargs...
+)
+  new_mts = copy(message_tensors(bpc))
+  c = 0
+  for edges in edge_groups
+    bpc_t, ct = update(bpc, edges; kwargs...)
+    for e in edges
+      new_mts[e] = message_tensor(bpc_t, e)
+    end
+    c += ct
+  end
+  return BeliefPropagationCache(copy(partitioned_itensornetwork(bpc)), new_mts),
+  c / (length(edge_groups))
+end
+
+"""
+More generic interface for update, with default params
+"""
+function update(
+  bpc::BeliefPropagationCache;
+  edges=default_edge_sequence(bpc),
+  niters=default_bp_niters(bpc),
+  target_precision=nothing,
+  verbose=false,
+  kwargs...,
+)
+  compute_norm = !isnothing(target_precision)
+  if isnothing(niters)
+    error("You need to specify a number of iterations for BP!")
+  end
+  for i in 1:niters
+    bpc, c = update(bpc, edges; compute_norm, kwargs...)
+    if compute_norm && c <= target_precision
+      if verbose
+        println("BP converged to desired precision after $i iterations.")
+      end
+      break
+    end
+  end
+  return bpc
+end
+
+"""
+Update the tensornetwork inside the cache
+"""
+function update(bpc::BeliefPropagationCache, states::Vector{ITensor}, vertices::Vector)
+  bpc = copy(bpc)
+  tn = tensornetwork(bpc)
+
+  for (state, vertex) in zip(states, vertices)
+    setindex_preserve_graph!(tn, state, vertex)
+  end
+  return bpc
+end
+
+function update(bpc, state, vertex)
+  return update(bpc, ITensor[state], [vertex])
+end
+
+"""
+Get the relevant message tensors coming on to the vertices (of the underlying itensornetwork in the cache)
+"""
+function environment_tensors(bpc::BeliefPropagationCache, verts::Vector)
```

Comment:
Maybe `incoming_messages`, as suggested above as part of the message update function.

Reply:
Yes I will rewrite it all as `incoming_messages` now